### PR TITLE
Miky

### PR DESCRIPTION
**Pull request by michaelhasty**
CR For Load Balancer Package.
Set up Load Balancer package using builder and strategy design patterns.
Consists of a load balancer interface, load balancer builder and implementation. Consists of a load balancer strategy abstract class and implementation of two load balancing strategies.

### CODE REVIEW

File: src/main/java/requestManagement/loadBalancer/LBRoundRobin.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import requestManagement.fleetManager.FleetManager;
+
+public class LBRoundRobin extends LBStrategy {
+
+    public LBRoundRobin(FleetManager fleetManager){
+        super(fleetManager);
+    }
+
+    @Override
+    public String getNextHost(){
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LBRoundRobin.java]: Add space between () and {

File: src/main/java/requestManagement/loadBalancer/LBStrategy.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package requestManagement.loadBalancer;
+
+import requestManagement.fleetManager.FleetManager;
+
+public abstract class LBStrategy {
+
+
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LBStrategy.java]: Remove whitespace

File: src/main/java/requestManagement/loadBalancer/LoadBalancer.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package requestManagement.loadBalancer;
+
+public interface LoadBalancer {
+
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LoadBalancer.java]: Remove whitespace

File: src/main/java/requestManagement/loadBalancer/LoadBalancer.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+    static LoadBalancerBuilder getBuilder(){
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LoadBalancer.java]: Add whitespace between () { (Do for all of these)

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package requestManagement.loadBalancer;
+
+import requestManagement.fleetManager.FleetManager;
+
+public class LoadBalancerBuilder {
+
+    private LBStrategy loadBalancingStrategy;
+    private FleetManager fleetManager;
+
+    LoadBalancerBuilder() {
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Make one line

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package requestManagement.loadBalancer;
+
+import requestManagement.fleetManager.FleetManager;
+
+public class LBClientSideRandom extends LBStrategy {
+
+    public LBClientSideRandom(FleetManager fleetManager) {
+        super(fleetManager);
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: Styling error -  change to single line method

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package requestManagement.loadBalancer;
+
+import requestManagement.fleetManager.FleetManager;
+
+public class LoadBalancerImpl  implements LoadBalancer {
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: Remove extra space

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    }
+
+    @Override
+    public String getNextHost() {
+        return "Client Side";
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: Styling error -  change to single line method

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    private LBStrategy loadBalancingStrategy;
+    private FleetManager fleetManager;
+
+    LoadBalancerImpl(LoadBalancerBuilder builder) {
+        loadBalancingStrategy = builder.getLoadBalancingStrategy();
+        fleetManager = builder.getFleetManager();
+    }
+
+    public String executeStrategy() { return loadBalancingStrategy.getNextHost(); }
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: Expand

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    public LBStrategy getLoadBalancingStrategy() {
+        return loadBalancingStrategy;
+    }
+
+    public FleetManager getFleetManager() {
+        return fleetManager;
+    }
+
```
Comment (Cribbin) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: Remove whitespace

File: src/main/java/requestManagement/loadBalancer/LBRoundRobin.java (hunk 1 above)
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LBRoundRobin.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LBRoundRobin.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package requestManagement.loadBalancer;
+
+import requestManagement.fleetManager.FleetManager;
+
+public class LBRoundRobin extends LBStrategy {
+
+    public LBRoundRobin(FleetManager fleetManager){
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LBRoundRobin.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LBStrategy.java (hunk 2 above)
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LBStrategy.java]: Whitespace

File: src/main/java/requestManagement/loadBalancer/LoadBalancer.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package requestManagement.loadBalancer;
+
+public interface LoadBalancer {
+
+
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancer.java]: Whitespace

File: src/main/java/requestManagement/loadBalancer/LoadBalancer.java (hunk 4 above)
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancer.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java (hunk 5 above)
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Single line please

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public LoadBalancerBuilder withLoadBalancingStrategy(LBStrategy loadBalancingStrategy) {
+        this.setLoadBalancingStrategy(loadBalancingStrategy);
+
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Whitespace

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        return this;
+    }
+
+    public LoadBalancerBuilder withFleetManager(FleetManager fleetManager) {
+        this.setFleetManager(fleetManager);
+
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Whitespace

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        return this;
+    }
+
+    private void setLoadBalancingStrategy(LBStrategy loadBalancingStrategy) {
+        this.loadBalancingStrategy = loadBalancingStrategy;
+    }
+
+    private void setFleetManager(FleetManager fleetManager){
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        this.fleetManager = fleetManager;
+    }
+
+    LBStrategy getLoadBalancingStrategy(){
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        return loadBalancingStrategy;
+    }
+
+    FleetManager getFleetManager(){
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        return fleetManager;
+    }
+
+    public LoadBalancerImpl build(){
```
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerBuilder.java]: Space between () and {

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java (hunk 10 above)
Comment (Tom-J-Barrett) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: Whitespace

File: src/main/test/LoadBalancerTest.java [hunk 19]
```diff
@@ -0,0 +1,20 @@
+import requestManagement.fleetManager.FleetManager;
+import requestManagement.loadBalancer.LBRoundRobin;
+import requestManagement.loadBalancer.LBStrategy;
+import requestManagement.loadBalancer.LoadBalancer;
+import org.junit.jupiter.api.Test;
+
```
Comment (Tom-J-Barrett) [inline src/main/test/LoadBalancerTest.java]: Whitespace

File: src/main/test/TestRequestManagement.java [hunk 20]
```diff
@@ -4,7 +4,8 @@
 import requestManagement.RequestManager;
 import requestManagement.Service;
 import requestManagement.fleetManager.FleetManager;
-import requestManagement.loadBalancing.LoadBalancer;
+import requestManagement.loadBalancer.LoadBalancer;
+import requestManagement.loadBalancer.LoadBalancerImpl;
```
Comment (Tom-J-Barrett) [inline src/main/test/TestRequestManagement.java]: Whitespace

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java (hunk 6 above)
Comment (Cribbin) [reply src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: I vote to keep it as it

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java (hunk 8 above)
Comment (Cribbin) [reply src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: Same as above

File: src/main/java/requestManagement/loadBalancer/LBRoundRobin.java [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+import requestManagement.fleetManager.FleetManager;
+
+public class LBRoundRobin extends LBStrategy {
+
+    public LBRoundRobin(FleetManager fleetManager) {
+        super(fleetManager);
+    }
+
+    @Override
+    public String getNextHost(){
```
Comment (RobHoodless) [inline src/main/java/requestManagement/loadBalancer/LBRoundRobin.java]: sp / LBRoundRobin / LbRoundRobin

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+import requestManagement.fleetManager.FleetManager;
+
+public class LBClientSideRandom extends LBStrategy {
+
+    public LBClientSideRandom(FleetManager fleetManager) {
+        super(fleetManager);
+    }
+
+    @Override
+    public String getNextHost() {
```
Comment (RobHoodless) [inline src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: You can change this method to return an ActiveHost

File: src/main/java/requestManagement/loadBalancer/LBStrategy.java [hunk 23]
```diff
@@ -0,0 +1,10 @@
+package requestManagement.loadBalancer;
+
+import requestManagement.fleetManager.FleetManager;
+
+public abstract class LBStrategy {
```
Comment (RobHoodless) [inline src/main/java/requestManagement/loadBalancer/LBStrategy.java]: sp / LBStrategy / LbStrategy

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+public class LoadBalancerImpl implements LoadBalancer {
+    private LBStrategy loadBalancingStrategy;
+    private FleetManager fleetManager;
+
+    LoadBalancerImpl(LoadBalancerBuilder builder) {
+        loadBalancingStrategy = builder.getLoadBalancingStrategy();
+        fleetManager = builder.getFleetManager();
+    }
+
+    public String executeStrategy() {
```
Comment (RobHoodless) [inline src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: I don't think this method needs to be externally visible. 

The primary external point of the Load Balancer object should be to accept a Http Request and return a Http Response. The process of choosing a host and executing the Http Request should be encapsulated.

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java (hunk 6 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: Left unchanged.

File: src/main/java/requestManagement/loadBalancer/LBRoundRobin.java (hunk 1 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LBRoundRobin.java]: Fixed

File: src/main/java/requestManagement/loadBalancer/LoadBalancer.java (hunk 3 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LoadBalancer.java]: Fixed

File: src/main/java/requestManagement/loadBalancer/LoadBalancer.java (hunk 4 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LoadBalancer.java]: Fixed for all

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java (hunk 7 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: removed

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java (hunk 9 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: expanded

File: src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java (hunk 10 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LoadBalancerImpl.java]: expanded

File: src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java (hunk 8 above)
Comment (michaelhasty) [reply src/main/java/requestManagement/loadBalancer/LBClientSideRandom.java]: Left unchanged

**Pull request merged**